### ISSUE TITLE
RFC: incoming message handlers

### ISSUE THREAD

**Issue by lperson** [ProgressiveCoders/Spoke#1317]
## Proposal:
Build a lightweight framework that allows adding incoming message processors.  An incoming message processor allows a Spoke deployment to add optional processing for each incoming message.

## Motivation:
There are at least three use cases for this that I am aware of.
* **Profanity filter** -- if a message from a contact contains profanity or abusive language, reassign the conversation to an administrator or apply a tag. See #359.
* **Automatic opt out** -- if a message from a contact contains the word STOP, opt out the contact. See #1016.
* **Automatic tagging** -- if a message appears to be in Spanish, apply a tag (for example).

## Detailed Design:
* Allow a list of incoming message handlers to be specified in the MESSAGE_HANDLERS environment variable.  If that variable is not specified it will default to `['save-message']`.
* `save-message` is the existing functionality in `src/server/api/lib/message-sending.js saveIncomingMessage`.
* The server will dynamically load each of the specified handlers, which must be placed in the `src/server/message_handlers` directory.  
* File names must match the names of the handlers  specified in MESSAGE_HANDLERS with a `.js` extension.  
* Each file must export a function called `processMessage` with takes the new incoming message instance and returns the message instance or something falsey.  
* If a handler returns something falsey, no other handles are called.  
* A handler can return a modified copy of the new incoming message if its function is to do some transformation on incoming messages.

## Additional Context:
A handler can do things other than returning something falsey or transforming an incoming message.  For example, a handler could reassign the contact associated with the incoming message, or a handler could apply a tag to an incoming message.

**Comment by ibrand** [comment]
Tagging some core contributors who might have input on this:
@schuyler1d @azuzunaga @filafb 

Overall, I'm really liking the sound of this. I'm not sure what other types of options we've explored for things like the profanity filter, but I like how this abstracts that process out

**Comment by azuzunaga** [comment]
Thanks for the tag, @ibrand!

I really like this idea too. I think the most important part of this design would be to specify and limit the number of side effects for the handlers. I can imagine if this isn’t kept tight we could end up dealing with some tricky bugs.

Thanks @lperson for putting this together!

**Comment by lperson** [comment]
Yeah. Definitely need to wrap the call to each handler in a try catch.

Unfortunately since the handlers are in process they can use any library in
our code and do whatever they want. I don't know how to prevent that. Any
suggestions?

**Comment by lperson** [comment]
@azuzunaga In the previous comment I noted that if nothing else changes, we should wrap the call to the handler in a try catch (which I did not do).

Action handlers are a precedent in Spoke for loading arbitrary modules and running functions in them. Look at this line in `/src/server/api/schema.js`  (ca. line 985)

```
const handler = require(`../action_handlers/${interactionStepAction}.js`);
```

Incoming message handlers would be created by people with access to a server who theoretically are interested in and accountable for keeping their Spoke running smoothly.  Some may be merged into Spoke itself, and those would be subject to code review and testing.  Organizations that deploy incoming message handlers would need to know they're doing so at their own risk, and if they're smart they'll have their code reviewed by someone who is familiar with Spoke.

However, I have been thinking about ways to isolate arbitrary code if we want to do that, and I have a couple of ideas.  

1. Implement handlers as microservices.  The framework would use an http library to call into the service.  We could ship Spoke with a process to host code for such microservices, or organizations could expose them separately.  This would force action handlers to use the API to mutate database state and it would provide process isolation.

2. Run the handlers in a vm, which would allow us to control what they have access to.  Again, we could require that they use the API, or we could pass libraries into them.  However, if we prevent them from using `require` at all, then they couldn't load modules they may need for their implementation.

I'm interested in your thoughts.

**Comment by ibrand** [comment]
@lperson and @azuzunaga I think microservices and VMs might be overkill, but would be very open to someone feeling differently. Something I like about the way action_handlers are working right now is that there's a test file that more or less serves as an "interface" even though it doesn't do validation and is really just a template that lives here: `src/server/action_handlers/test-action.js`

I've created #1321 and assigned it to myself to add more documentation around the action handler functionality and could totally create a similar doc for message handlers once they exist

**Comment by azuzunaga** [comment]
Thanks @ibrand and @lperson! I tend to agree with Ilona on microservices being overkill (although if some workloads are demanding enough it might be worth doing). I think the best compromise would be to have a few guidelines to follow for both authors and reviewers. One thing I would suggest is requiring that each handler accept as arguments all dependencies. I think this would make the handlers easier to test and to reason about.

**Comment by Frydafly** [comment]
Checking in -- this was merged in v7.0?

**Comment by schuyler1d** [comment]
Yes! (though I should note the Detailed Design section while mostly supported in features isn't precise in approach -- see documentation for Message Handlers for APIs)
